poetry/core/packages/package_specification.py

```diff
@@ -0,0 +1,118 @@
+from typing import FrozenSet
+from typing import List
+from typing import Optional
+
+from poetry.core.utils.helpers import canonicalize_name
+
+
+class PackageSpecification(object):
+    def __init__(
+        self,
+        name,
+        source_type=None,
+        source_url=None,
+        source_reference=None,
+        source_resolved_reference=None,
```

Comment:
It would be good for these to move into an object (and later to a dataclass once we drop 2.7). It becomes easier to maintain.

Reply:
After the release of the `1.2` version of Poetry I plan on refactoring this bit by using sources objects that will be attached to dependency objects.

This will remove the need to have multiple dependency classes.